src/k8s/pkg/client/kubernetes/upgrades.go

```
@@ -0,0 +1,169 @@
+package kubernetes
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"slices"
+
+	"github.com/canonical/k8s/pkg/log"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+// TODO: If the upgrade CRD grows, consider using kubebuilder.
+type Upgrade struct {
+	Metadata struct {
+		Name string `json:"name"`
+	} `json:"metadata"`
+	Status struct {
+		Phase         string   `json:"phase"`
+		UpgradedNodes []string `json:"upgradedNodes"`
+	} `json:"status"`
+}
+
+func (c *Client) K8sdIoRestClient() (*rest.RESTClient, error) {
+	k8sdConfig := c.RESTConfig()
+	k8sdConfig.GroupVersion = &schema.GroupVersion{Group: "k8sd.io", Version: "v1alpha"}
+	k8sdConfig.NegotiatedSerializer = serializer.NewCodecFactory(runtime.NewScheme())
+
+	restClient, err := rest.RESTClientFor(k8sdConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create REST client for k8sd.io group: %w", err)
+	}
+	return restClient, nil
+}
+
+func (c *Client) GetInProgressUpgrade(ctx context.Context) (*Upgrade, error) {
+	log := log.FromContext(ctx).WithValues("upgrades", "GetInProgressUpgrade")
+
+	restClient, err := c.K8sdIoRestClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create REST client for k8sd.io group: %w", err)
+	}
+
+	upgrades, err := restClient.Get().AbsPath("/apis/k8sd.io/v1alpha/upgrades").DoRaw(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get upgrades: %w", err)
+	}
+	log.Info("Got upgrades", "upgrades", string(upgrades))
+	var result struct {
+		Items []Upgrade `json:"items"`
+	}
+	if err := json.Unmarshal(upgrades, &result); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal upgrades: %w", err)
+	}
+
+	for _, upgrade := range result.Items {
+		if upgrade.Status.Phase != "Failed" && upgrade.Status.Phase != "Completed" {
+			return &upgrade, nil
+		}
+	}
```

if there are multiple upgrades, this picks the first one and silently ignores the rest. is this intended?

---

@eaudetcobello normally we should only have one update but you raised a good point.
I think it is the best solution to:

log the issue and upgrades
sort by name (naming is cluster-upgrade-to-rev-<rev>)
return the last entry.
This will always return the latest upgrade even if there is a stale upgrade left for some reason.